### ISSUE TITLE
Simple keras model, Model.fit() does not learn unless experimental_run_tf_function=False at compile

### ISSUE THREAD

**Issue by fcarsten** [tensorflow/tensorflow#36044]
**System information**
- Windos 10, 64 Bit
- TensorFlow installed from pip
- TensorFlow version: v2.1.0-rc2-17-ge5bf8de410 2.1.0
- Python version: 3.7

**Describe the current behavior**

Simple Model (see below), compiles without error. It is used in a reinforcement scenario, i.e. iterations of `predict()` and `fit()` calls to iteratively train the model. 
Currently the model does not seem to improve with calls to `fit()` unless `compile()` was called with `compile(..., experimental_run_tf_function = False)`. 

**Describe the expected behavior**

Model should train equally well whether `experimental_run_tf_function = False` was passed to `model.compile()` or not.

**Code to reproduce the issue**

Example code can be found at [https://github.com/fcarsten/tic-tac-toe/blob/tf-2.1-issue/test_nn_q_tf2.py]( https://github.com/fcarsten/tic-tac-toe/blob/tf-2.1-issue/test_nn_q_tf2.py) - need to check out the whole branch as it uses other code from this repository. 

Either run as is `run_test(True)` to see failing, or `run_test(False)` to see how it should run. 

When running as expected "Player 1 win %" should increase and end up over 80%, usually somewhere around 90%. When not running as expected, "Player 1 win %" will randomly meander up and down.

The model is defined in File SimpleNNQPlayerTF2.py line 29 ff:
```
input_layer = tf.keras.Input(shape=(BOARD_SIZE * 3,))
x = tf.keras.layers.Dense(BOARD_SIZE * 3 * 9, activation='relu')(input_layer)
x = tf.keras.layers.Dense(BOARD_SIZE * 3 * 100, activation='relu')(x)
x = tf.keras.layers.Dense(BOARD_SIZE * 3 * 9, activation='relu')(x)
q_values = tf.keras.layers.Dense(BOARD_SIZE, activation=None, name='q_values')(x)
probabilities = tf.keras.layers.Softmax(name='probabilities')(q_values)

self.model = tf.keras.Model(inputs=input_layer, outputs=[probabilities, q_values])
if run_tf_function:
    self.model.compile(optimizer='adam', loss = [None, tf.keras.losses.MeanSquaredError()])
else:
    self.model.compile(optimizer='adam', loss = [None, tf.keras.losses.MeanSquaredError()], experimental_run_tf_function = False)
```

**Other info / logs**

While the model is a very simple sequential model, note that it has 2 output layers and the training target layer is not the final layer in the model. Not sure this makes any difference,

**Comment by ravikyram** [comment]
@fcarsten 

Looks like code is incomplete. Request you to share colab link or simple stand alone code to reproduce the issue in our environment. It helps us in localizing the issue faster. Thanks!

**Comment by fcarsten** [comment]
Colab link as requested: [https://colab.research.google.com/drive/1rS0RLOL7zMh3hDC4bAwAPGjI6qCXdNEy](https://colab.research.google.com/drive/1rS0RLOL7zMh3hDC4bAwAPGjI6qCXdNEy)

**Comment by jvishnuvardhan** [comment]
@fcarsten Not sure what is the root-cause of the issue. I changed order of final two lines of your code from Sequence #1: `run_test(True) then run_test(False)` to Sequence #2: `run_test(False) then run_test(True). The results are very different than what you noticed with the Sequence #1. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/04c083b0839571ce2c80f41949256254/tensorflow-issue-36044.ipynb). Thanks!

**Comment by fcarsten** [comment]
@jvishnuvardhan Looking at your gist, it looks like you are using a different version of TF than me. 

Yours is `2.2.0-dev20200121` whereas mine was `v2.1.0-rc2-17-ge5bf8de410 2.1.0`. 

Looking at the latest source code it seems that the work-around flag I use to make it work is no longer supported in `2.2.0-dev20200121`, so it is always failing now.


v2.1.0-rc2-17-ge5bf8de410 2.1.0:
training.py#compile():
```
...
303:    self._experimental_run_tf_function = kwargs.pop(
             'experimental_run_tf_function', True)
...
```

no longer exists in 2.2.0-dev20200121.

So, the problem is still there, just my workaround no longer works. (which from my point of view is kind of worse ;-) )
...

**Comment by fcarsten** [comment]
Running `fit()` with `verbose=1` shows the difference in behavior even more.

With `experimental_run_tf_function = False` the loss decreases slowly but consistently. Without however the loss actually increases until it stabilizes around 400-500:

```
Train on 3 samples

3/3 [==============================] - 0s 71ms/sample - loss: 0.0484 - q_values_loss: 0.0484
Train on 4 samples

4/4 [==============================] - 0s 3ms/sample - loss: 0.1187 - q_values_loss: 0.1187
Train on 5 samples

5/5 [==============================] - 0s 2ms/sample - loss: 0.3707 - q_values_loss: 0.3707
Train on 3 samples

3/3 [==============================] - 0s 3ms/sample - loss: 1.4076 - q_values_loss: 1.4076
Train on 3 samples

3/3 [==============================] - 0s 3ms/sample - loss: 2.3861 - q_values_loss: 2.3861
Train on 5 samples

5/5 [==============================] - 0s 2ms/sample - loss: 4.5628 - q_values_loss: 4.5628
Train on 3 samples

3/3 [==============================] - 0s 3ms/sample - loss: 10.1721 - q_values_loss: 10.1721
Train on 3 samples

3/3 [==============================] - 0s 3ms/sample - loss: 17.7317 - q_values_loss: 17.7317
Train on 4 samples

4/4 [==============================] - 0s 2ms/sample - loss: 27.7600 - q_values_loss: 27.7600
Train on 5 samples

5/5 [==============================] - 0s 2ms/sample - loss: 35.3489 - q_values_loss: 35.3489
Train on 3 samples

3/3 [==============================] - 0s 4ms/sample - loss: 50.9869 - q_values_loss: 50.9869
Train on 3 samples

3/3 [==============================] - 0s 3ms/sample - loss: 58.2381 - q_values_loss: 58.2381
Train on 5 samples

[...]

3/3 [==============================] - 0s 3ms/sample - loss: 310.6920 - q_values_loss: 310.6920
Train on 5 samples

5/5 [==============================] - 0s 2ms/sample - loss: 323.8771 - q_values_loss: 323.8771
Train on 5 samples

5/5 [==============================] - 0s 2ms/sample - loss: 341.6482 - q_values_loss: 341.6482
Train on 5 samples

5/5 [==============================] - 0s 2ms/sample - loss: 338.4795 - q_values_loss: 338.4795
Train on 3 samples

3/3 [==============================] - 0s 3ms/sample - loss: 321.3784 - q_values_loss: 321.3784
Train on 4 samples

4/4 [==============================] - 0s 2ms/sample - loss: 334.4857 - q_values_loss: 334.4857
Train on 3 samples

3/3 [==============================] - 0s 3ms/sample - loss: 410.9324 - q_values_loss: 410.9324
Train on 3 samples

3/3 [==============================] - 0s 3ms/sample - loss: 394.3834 - q_values_loss: 394.3834
Train on 4 samples

4/4 [==============================] - 0s 2ms/sample - loss: 410.5582 - q_values_loss: 410.5582
Train on 4 samples

4/4 [==============================] - 0s 2ms/sample - loss: 412.2753 - q_values_loss: 412.2753
Train on 5 samples

5/5 [==============================] - 0s 2ms/sample - loss: 410.8856 - q_values_loss: 410.8856
```

**Comment by fcarsten** [comment]
@pavithrasv I think I found a way to show the different behaviour in TensorFlow nightly by using `with tf.Graph().as_default():` instead of the `experimental_run_tf_function` argument:

[https://colab.research.google.com/drive/1clvvsr9i0Tvt8o3vyy8N3FSdSdX5-Se6](https://colab.research.google.com/drive/1clvvsr9i0Tvt8o3vyy8N3FSdSdX5-Se6)

I also added a TensorBoard visualization of the loss behaviour. Orange is with eager on and Blue with eager off (both on log scale):

![image](https://user-images.githubusercontent.com/554786/73318752-7d1c0900-428e-11ea-8e61-00d1ffb566e6.png)

Note: Both runs have the same number of iterations, the Blue line is only shorter because execution time is much faster with eager off.

**Comment by fcarsten** [comment]
I think I narrowed the problem further down. The current code is basically:
```
input_layer = tf.keras.Input(shape=(BOARD_SIZE * 3,))
x = tf.keras.layers.Dense(BOARD_SIZE * 3 * 9, activation='relu')(input_layer)
x = tf.keras.layers.Dense(BOARD_SIZE * 3 * 100, activation='relu')(x)
x = tf.keras.layers.Dense(BOARD_SIZE * 3 * 9, activation='relu')(x)
q_values = tf.keras.layers.Dense(BOARD_SIZE, activation=None, name='q_values')(x)
probabilities = tf.keras.layers.Softmax(name='probabilities')(q_values)

self.model = tf.keras.Model(inputs=input_layer, outputs=[probabilities, q_values])
self.model.compile(optimizer='adam', loss = [None, tf.keras.losses.MeanSquaredError()])
```

and training happens with:
```
res = self.model.fit(np_inputs, {'q_values': np_targets}, verbose=0)
```

This works with eager mode off, but does not work with eager mode enabled.

If I change the code slightly by changing the order of the ouputs and loss arguments:

```
input_layer = tf.keras.Input(shape=(BOARD_SIZE * 3,))
x = tf.keras.layers.Dense(BOARD_SIZE * 3 * 9, activation='relu')(input_layer)
x = tf.keras.layers.Dense(BOARD_SIZE * 3 * 100, activation='relu')(x)
x = tf.keras.layers.Dense(BOARD_SIZE * 3 * 9, activation='relu')(x)
q_values = tf.keras.layers.Dense(BOARD_SIZE, activation=None, name='q_values')(x)
probabilities = tf.keras.layers.Softmax(name='probabilities')(q_values)

self.model = tf.keras.Model(inputs=input_layer, outputs=[q_values, probabilities])
self.model.compile(optimizer='adam', loss = [tf.keras.losses.MeanSquaredError(), None])
```

this now seems to work in both version (although still much slower with eager mode enabled).

One potential explanation might be that the order of `loss` functions with eager mode disabled is the same as the order of the specified `outputs` in the `Model` constructor, but for some reason with eager mode enabled the order of `loss` functions is matching the order they appear in the graph. Not sure if that's the case, but it would explain the observed behaviour.

**Comment by fcarsten** [comment]
@pavithrasv I dug around a bit more and I think this is the point where things go pear-shaped:

training_eager.py#_model_loss(), currently around line 147:
```
[...]
    loss_fns = [
        loss_fn for loss_fn in model.loss_functions if loss_fn is not None
    ]
[...]
```
this removes `None` entries from the list of loss functions. However the corresponding outputs are not removed from the `outs` list. This results in the lists becoming un-aligned and all subsequent loss computations using the wrong outputs to apply the loss function to.

**Comment by pavithrasv** [comment]
Thank you @fcarsten for taking a look. We are doing a bunch of code refactoring internally and i think this use case in eager will be fixed as part of that. Will update this thread after I know when it is fixed in a tf-nightly release.

**Comment by jvishnuvardhan** [comment]
@fcarsten Looks like this was resolved. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/412bf7495d33e61fde09ee6736c01d83/tensorflow-issue-36044.ipynb). 

Please verify once and close the issue if this was resolved for you. Thanks!

**Comment by fcarsten** [comment]
Thanks, looks good. I'm snowed under at the moment, so can't immediately validate but happy to close it based on @jvishnuvardhan results. Can always reopen if evidence to the contrary emerges.
